src/main/java/hudson/plugins/parameterizedtrigger/BuildTrigger.java

```diff
@@ -50,18 +47,39 @@ public BuildStepMonitor getRequiredMonitorService() {
 	@Override @SuppressWarnings("deprecation")
 	public boolean perform(AbstractBuild<?, ?> build, Launcher launcher,
 			BuildListener listener) throws InterruptedException, IOException {
+
+		HashSet<BuildTriggerConfig> alreadyFired = new HashSet<BuildTriggerConfig>();
+
+		// If this project has non-abstract projects, we need to fire them
+		for (BuildTriggerConfig config : configs) {
+			boolean hasNonAbstractProject = false;
+
+			List<Job> jobs = config.getProjectList(build.getRootBuild().getProject().getParent(), build.getEnvironment(listener));
+			for (Job j : jobs) {
+				if (!(j instanceof AbstractProject)) {
+					hasNonAbstractProject = true;
+				}
+			}
+			// Fire this config's projects if not already fired
+			if (hasNonAbstractProject) {
+				config.perform(build, launcher, listener);
+				alreadyFired.add(config);
+			}
+		}
+
         if (canDeclare(build.getProject())) {
             // job will get triggered by dependency graph, so we have to capture buildEnvironment NOW before
             // hudson.model.AbstractBuild.AbstractBuildExecution#cleanUp is called and reset
             EnvVars env = build.getEnvironment(listener);
             build.addAction(new CapturedEnvironmentAction(env));
-        } else {
+        } else {  // Not using dependency graph
             for (BuildTriggerConfig config : configs) {
-                config.perform(build, launcher, listener);
+				if (!alreadyFired.contains(config)) {
```

Comment:
@jtnord Fair enough, it's rather confusing!  

The upstream project is always an AbstractProject (you can't add the trigger on a workflow job, but workflow has its own mechanism to fire downstream builds).  canDeclare is invoked on the *upstream* project, an AbstractProject, simply to see if the *upstream* can participate in the DependencyGraph, which might be true or false.  

Regardless of this, we need to ensure that workflow jobs are triggered.    I'm not going to guarantee there aren't edge cases until the tests are reviewed, but this part at least should work.

I hope that makes the rationale a bit clearer?  I'd be happy to do a hangout if it still seems iffy to you.